drivers: ramdisk: Add support for external initrd

Deedone opened the pull request:
Add support for using initrd provided by the bootloader. Supports both compressed and uncompressed filesystem images.

Deedone committed:
drivers: ramdisk: Add support for external initrd

Add support for using initrd provided by the bootloader as the RAM Disk.
Location of the initrd is specified by the DISK_RAM_START Kconfig option.
Currently only uncompressed filesystem images are supported.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>

xakep-amatop commented on drivers/disk/Kconfig.ram:
```diff
@@ -22,6 +22,19 @@ config DISK_RAM_VOLUME_NAME
 
+config DISK_RAM_EXTERNAL
+	bool "Use initrd provided by the bootloader"
+	default n
```
There isn't any sense in having it as `default n`, so just delete this line.

Please, read this for more information: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults

Deedone replied:
Done

xakep-amatop commented on drivers/disk/ramdisk.c:
```diff
@@ -33,6 +37,21 @@ static int disk_ram_access_status(struct disk_info *disk)
+	mm_reg_t virt_start;
+
+	if (!ramdisk_buf) {
+#if defined(CONFIG_MMU) || defined(CONFIG_PCIE)
```
why can't you use `DEVICE_MMIO_IS_IN_RAM` here?

Deedone replied:
Switched to only checking MMU config as PCIE is not relevant here

xakep-amatop commented on drivers/disk/ramdisk.c:
```diff
@@ -33,6 +37,21 @@ static int disk_ram_access_status(struct disk_info *disk)
+	if (!ramdisk_buf) {
+#if defined(CONFIG_MMU) || defined(CONFIG_PCIE)
+		device_map(&virt_start, CONFIG_DISK_RAM_START,
+					CONFIG_DISK_RAM_VOLUME_SIZE, K_MEM_CACHE_NONE);
```
Should we make this memory cacheable, or at least give the user the option to decide?

Deedone replied:
Done

xakep-amatop commented on drivers/disk/ramdisk.c:
```diff
@@ -33,6 +37,21 @@ static int disk_ram_access_status(struct disk_info *disk)
+
+	if (!ramdisk_buf) {
+#if defined(CONFIG_MMU) || defined(CONFIG_PCIE)
+		device_map(&virt_start, CONFIG_DISK_RAM_START,
```
Why can't you put `ramdisk_buf` here with explicit casting for suppressing warnings and reduce the diff?

xakep-amatop commented on drivers/disk/ramdisk.c:
```diff
@@ -33,6 +37,21 @@ static int disk_ram_access_status(struct disk_info *disk)
 static int disk_ram_access_init(struct disk_info *disk)
 {
+#if IS_ENABLED(CONFIG_DISK_RAM_EXTERNAL)
+	mm_reg_t virt_start;
```
Can we reduce the lifetime and visibility of this variable by moving its declaration under the if statement below?
looks like it produces warnings of unused variable in case when we have `CONFIG_DISK_RAM_EXTERNAL` enabled and `DEVICE_MMIO_IS_IN_RAM` is disabled

Deedone replied:
Done

xakep-amatop commented on drivers/disk/ramdisk.c:
```diff
@@ -33,6 +37,21 @@ static int disk_ram_access_status(struct disk_info *disk)
+
+	if (!ramdisk_buf) {
+#if defined(CONFIG_MMU) || defined(CONFIG_PCIE)
+		device_map(&virt_start, CONFIG_DISK_RAM_START,
```
Just out of curiosity, is there any reason for placing this call here instead of in disk_ram_init?

Deedone replied:
Moved it, thanks for suggestion

xakep-amatop commented:
Review finished, generally ok.
I have a few cosmetic suggestions and some questions

Deedone committed:
modules: lz4: Add lz4frame to the build system

Add lz4frame and xxhash to the build system. This is required to
support the decompression of LZ4 files made with LZ4 cli tool.
Base lz4 library only supports decompression of raw LZ4 blocks without
taking into account the frame metadata.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>

Deedone committed:
drivers: ramdisk: Add support for compressed initrd

Add support for decompressing the initrd before using it as the RAM Disk.
For now, only LZ4 compression is supported.
It is advised to use smaller block sizes for the compressed initrd to
reduce the size of heap memory required for decompression.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>

firscity commented on drivers/disk/ramdisk.c:
```diff
@@ -12,13 +12,19 @@
 #include <zephyr/init.h>
 #include <zephyr/device.h>
 #include <zephyr/logging/log.h>
+#include "lz4frame.h"
```
Should we include it unconditionally to this file? Also should it be `#include ""` instead of `#include <>`?

Deedone replied:
Regarding " vs < it was included with " in a sample for some reason so I assumed that it is required, changed to <

firscity reviewed (commented):
LGTM with one comment

xakep-amatop commented:
Are we going to merge this PR?
